datafusion/common/src/tree_node.rs

```diff
@@ -88,132 +177,138 @@ pub trait TreeNode: Sized {
     ///
     /// If an Err result is returned, recursion is stopped immediately
     ///
-    /// If [`VisitRecursion::Stop`] is returned on a call to pre_visit, no
+    /// If [`TreeNodeRecursion::Stop`] is returned on a call to pre_visit, no
     /// children of that node will be visited, nor is post_visit
     /// called on that node. Details see [`TreeNodeVisitor`]
     ///
-    /// If using the default [`TreeNodeVisitor::post_visit`] that does
+    /// If using the default [`TreeNodeVisitor::f_up`] that does
     /// nothing, [`Self::apply`] should be preferred.
-    fn visit<V: TreeNodeVisitor<N = Self>>(
+    fn visit<V: TreeNodeVisitor<Node = Self>>(
         &self,
         visitor: &mut V,
-    ) -> Result<VisitRecursion> {
-        handle_tree_recursion!(visitor.pre_visit(self)?);
-        handle_tree_recursion!(self.apply_children(&mut |node| node.visit(visitor))?);
-        visitor.post_visit(self)
+    ) -> Result<TreeNodeRecursion> {
+        handle_visit_recursion_down!(visitor.f_down(self)?);
+        handle_visit_recursion_up!(self.apply_children(&mut |n| n.visit(visitor))?);
+        visitor.f_up(self)
```

Comment:
```suggestion
   match visitor.f_down(self)? {
            TreeNodeRecursion::Continue => {
                handle_visit_recursion_up!(
                    self.apply_children(&mut |n| n.visit(visitor))?
                );
                visitor.f_up(self)
            }
            TreeNodeRecursion::Jump => visitor.f_up(self),
            TreeNodeRecursion::Stop => Ok(TreeNodeRecursion::Stop),
        }
```

I have tried to change here like this to add  "f_up(a)" after  "f_down(a)" in the "f_down_jump_on_a_visits". "f_up(a)" is added but some other tests fail. It happens because as you have also said that previous skip behavior was in that way. What do you think that the failed test rules can be converted easily to work in this way? I know there is no exact truth and I am not sure which is more intuitive. However, I don't see any problem in keeping it like this.

Reply:
This is the https://github.com/apache/arrow-datafusion/pull/8891#discussion_r1490932349 topic, that whether we should to call `f_up(n)` after `f_down(n)` if `f_down` returns `Jump`.

If you changed only the `TreeNodeVisitor` logic in your experiment then it `ExprIdentifierVisitor` must be the root cause of the failures as that's the only one returning `TreeNodeRecursion::Jump` in its `f_down`.
I think we could fix those failures if we duplicated the https://github.com/apache/arrow-datafusion/blob/ec86acbc1fbc0da1e0bec9ad066a5177ec586c96/datafusion/optimizer/src/common_subexpr_eliminate.rs#L669-L671 from its `f_down` to its `f_up` and return the original node unmodified if the check is true.
But I feel this just confirms my ponint in https://github.com/apache/arrow-datafusion/pull/8891#discussion_r1490932349 that we can do anything in `f_down` with a node and if we prune the node's subtree (return `Jump`) then it makes no sense to call `f_up` on it.